### PR TITLE
add loop unrolling optimization

### PR DESCRIPTION
**Pull request by WalterBright**
This does some modest unrolling - only once, and only if the loop bounds are fixed. The flow analysis was there, I just had to put it together.

### CODE REVIEW

**Comment by WalterBright** [comment]
The code:
```
int foo() {
    int result = 0;
    for (int i = 0; i < 10; i++)
        result += i;
    return result;
}
```
formerly generated:
```
                xor     EAX,EAX
                xor     ECX,ECX
L4:             add     EAX,ECX
                inc     ECX
                cmp     ECX,0Ah
                jb      L4
                ret
```
and now:
```
                xor     EDX,EDX
                xor     ECX,ECX
L4:             lea     EAX,[01h][ECX*2]
                add     EDX,EAX
                add     ECX,2
                cmp     ECX,0Ah
                jb      L4
                mov     EAX,EDX
                ret
```

File: src/ddmd/backend/gloop.c [hunk 1]
```diff
@@ -794,9 +795,31 @@ void loopopt()
...
+            ln = l->Lnext;
+            if (loopunroll(l))
+            {
+                compdfo();                      // compute depth-first order
+                blockinit();
+                compdom();
+                findloops(&startloop);          // recompute block info
+                doflow = TRUE;
+                if (ln)
+                {   ln = startloop;         // start over
```
Comment (andralex) [inline src/ddmd/backend/gloop.c]: since you have no code following braces in the code introduced elsewhere, this is jarring

**Comment by andralex** [comment]
Well the code that does repeated additions should be ideally optimized to `result = 45;` i.e. merge repeated additions of constants.

**Comment by WalterBright** [comment]
> Well the code that does repeated additions should be ideally optimized to result = 45; i.e. merge repeated additions of constants.

True, but that would be a different optimization.

File: src/ddmd/backend/gloop.c (hunk 1 above)
Comment (WalterBright) [inline src/ddmd/backend/gloop.c]: It's copy-pasted from line 715.

**Comment by UplinkCoder** [comment]
@WalterBright this optimization needs to controllable somehow.
because it can increase i-cache pressure.

File: src/ddmd/backend/gloop.c [hunk 2]
```diff
@@ -794,9 +795,31 @@ void loopopt()
     /* one first, thus moving LIs out as far as possible.       */
 
     doflow = TRUE;                      /* do flow analysis             */
+
+    if (go.mfoptim & MFtime)
+    {
+        cmes("Starting loop unrolling\n");
+        for (loop *l = startloop; l; l = ln)
```
Comment (andralex) [inline src/ddmd/backend/gloop.c]: There's hardly a more awful symbol than `l`. Just look at the rendering in this github PR - is that condition `l` or `1`?

File: src/ddmd/backend/gloop.c (hunk 1 above)
Comment (andralex) [inline src/ddmd/backend/gloop.c]: That doesn't make it any better!

File: src/ddmd/backend/gloop.c (hunk 2 above)
Comment (WalterBright) [inline src/ddmd/backend/gloop.c]: In general you're quite right, and I considered it. But it's used consistently throughout `gloop.c`, its meaning is clear from the context, and being typed as a pointer means the compiler won't let you conflate it with `1`.

**Comment by WalterBright** [comment]
> controllable somehow

Yes, which is why it only unrolls once and only for loops with less than 100 nodes.

File: src/ddmd/backend/gloop.c (hunk 2 above)
Comment (MartinNowak) [inline src/ddmd/backend/gloop.c]: If we're at naming `nl` for next loop is more obvious than `ln` which typically abbreviates line.

**Comment by MartinNowak** [comment]
Looking forward at the regressions ;), e.g. more usage of the `O(N^2)` optimization due to big unrolled loops.

**Comment by WalterBright** [comment]
> big unrolled loops

Only small loops are unrolled - there's a size check.

**Pull request merged**